### ISSUE TITLE
Implement supporter wall widget reminder banner notification

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#5538]
When users haven't placed the TwG "Supporter wall" widget (possibly because they're not aware of it), Site Kit should remind the users about that via a corresponding banner notification.

This is not a launch blocker, but would be nice to include right away or at least shortly after the initial launch of the module.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A new transient should be set at the time the TwG module setup is completed, with a timeout of one week (see below why).
* One week after the Thank with Google setup has been completed (based on the above transient not being present): _If_ the site has any widget area (which it typically does unless it's a full site editing theme) but does not yet have the TwG supporter wall widget placed in any widget area (and the banner hasn't been dismissed by the current user, see below), a new banner notification should show up on the Site Kit dashboard, reminding the user about the possibility to add it:
    * The new banner should follow [this Figma design](https://www.figma.com/file/t5zDpQ28Lo7ZyRcOpGdqgu/TwG?node-id=2343%3A10859).
    * The CTA button should take the user to the site's Widgets admin screen.
    * If the user clicks the "Maybe later" button, the notification should be permanently dismissed (for that user).
* A Storybook story for that banner notification should be added.

## Implementation Brief

* In `ThankWithGoogle.php`:
    * Implement the `Module_With_Activation` interface and define the interface's `on_activation()` method.
    * Within the new `ThankWithGoogle::on_activation()` method, use the `Transient::set()` method where the transient name could be `googlesitekit_thank_with_google_setup`, the value could be `true` or the current timestamp (doesn't matter) and the expiration should be one week.
    * Create a simple `GET:prompt-supporter-wall` datapoint which returns true if TwG is connected but the transient defined above has expired. 
* In `assets/js/modules/thank-with-google/datastore/supporter-wall.js`:
    * Add `promptSupporterWall` to the `baseInitialState` as `undefined`.
    * Use the `createFetchStore()` method to create a new simple store which calls the `GET:prompt-supporter-wall` endpoint. The `basename` should be `getPromptSupporterWall`. Add this store to the `Data.combineStores()` function which is then exported as the final store.
    * Similar to the existing `getSupporterWallSidebars` store, also add a `getPromptSupporterWall` generator function to `baseResolvers` that calls the `fetchGetPromptSupporterWall` action.
    * Similar to the existing `getSupporterWallSidebars` store, add another selector function to `baseSelectors` which returns the new `promptSupporterWall` state variable.
* In `assets/js/components/notifications`, create a new React component `ThankWithGooglePromptSupporterWall`. Within this new component:
    * Use the `displaySupporterWallPrompt()` selector created above - if it returns false, return `null` early.
    * Otherwise, render the `BannerNotification` component with the following props:
        * `id`: `twg-prompt-supporter-wall`
        * `title` and `description`: translatable strings copied from the Figma mock.
        * `ctaLabel`: Translatable "Add Supporter Wall widget" string as per Figma.
        * `ctaLink`: Use the `getWidgetsAdminURL()` selector on the `CORE_SITE` datastore to fetch the link to the widgets admin screen.
        * `dismiss`: Translatable "Maybe later" string as per Figma.
        *  `isDismissible`
* Render the `ThankWithGooglePromptSupporterWall` component within `BannerNotifications` above the `ZeroDataStateNotifications` banner. 
* Add a storybook story for the `ThankWithGooglePromptSupporterWall` notification banner.

### Test Coverage

* Add tests to the `supporter-wall.test.js` datastore tests.

## QA Brief

- Setup Thank with Google using a VPN that is connected to the US region.
- Do not add any supporter wall widgets yet.
- Ensure the banner notification doesn’t appear.
- Within the database -> `wp_options` table, verify the value of `_transient_timeout_googlesitekit_thank_with_google_setup` is a timestamp 7 days from now. Change this value of  `googlesitekit_thank_with_google_setup` from the `wp_options` table to the current timestamp. (This mimics invalidating the transient signalling the banner to show). Ideally, set a date 7 days in the future following the [QA process](https://github.com/google/site-kit-wp/issues/5272#issuecomment-1223996292) used in #5272 
- Refresh the SiteKit dashboard and ensure the new banner notification appears as per the AC.
- Ensure the `Add Supporter Wall widget` CTA link takes to the WP widgets page.
- Add a supporter wall widget and ensure the banner notification doesn’t appear.
- Remove the widget and ensure the notification appears.
- Ensure this [story](https://google.github.io/site-kit-wp/storybook/develop/?path=/story/components-thankwithgooglesupporterwallnotification--ready) satisfies the AC.

**Updated**
- Following the above steps, ensure the issue found in this [comment](https://github.com/google/site-kit-wp/issues/5538#issuecomment-1263322636) was fixed.
- Enable the TwG feature flag.
- Disable SSL and verify the above issue doesn't appear.
- Ensure the flickering notification issue found in this [comment](https://github.com/google/site-kit-wp/issues/5538#issuecomment-1262342041) is fixed
- Ensure the following React warning that doesn't appear when the notification is displayed.

![Screenshot 2022-10-04 at 9 49 13 PM](https://user-images.githubusercontent.com/24408261/193884376-6936bb91-e6aa-4d54-9c63-0b726bfa95bb.png)

## Changelog entry

* Add conditional reminder to set up the Supporter Wall widget for Thank with Google.

**Comment by tofumatt** [comment]
Approach in the IB looks good, but I read the ACs/intent of this issue as also _rendering_ the notification for users. We should add a note to the IB that we should add the newly-created notification to [`BannerNotifications`](https://github.com/google/site-kit-wp/blob/18b0dbde3fe44bbb7be794ed0183b398e3093846/assets/js/components/notifications/BannerNotifications.js#L60) so it appears for users.

**Comment by jimmymadon** [comment]
@hussain-t In addition to my CR comments, there is a slight issue here. The AC says 'new transient should be set at the time the TwG module **setup is completed**'. Right now, as per the IB, we add the transient when the module is `activated`. 

In my testing, I realised that if the user attempts to connect TwG, but does not continue the 'Customize' step and clicks on cancel, the module is 'activated' but the setup is not complete. The 7 day timer transient currently gets added at this point which seems to be incorrect. As if the user fully completes the TwG customize step later (say after 8 days), he instantly sees the 'Supporter Wall Notification'. We may have to add the transient after the customize step somehow.

**Comment by jimmymadon** [comment]
As per my previous comment, we cannot set the 7 day timer within the `onActivation()` method. 

We can create a possible `onConnection()` hook as follows: Within `ModuleSettings::merge()`, we can check if the module is connected before and after saving settings. If the module was found to have been connected only after the saving of settings, it indicates that that was the save which 'connected' the module and a `onConnection()` hook can be run at this time. However, this solution requires us to add the module instance within the ModuleSettings class which isn't ideal.

The only other way is to set the transient timer as a separate action dispatched from the front-end in `SetupCustomize` only if the module was connected for the first time.

**Comment by hussain-t** [comment]
Thanks for bringing this up, @jimmymadon.

>The only other way is to set the transient timer as a separate action dispatched from the front-end in SetupCustomize only if the module was connected for the first time.

According to your suggestion, how do we know if the user fires the save settings action for the first time from the frontend?

**Comment by techanvil** [comment]
@jimmymadon @hussain-t this is an interesting one. I have a suggestion which builds upon the idea of checking if the module is connected after saving the settings.

We could use WP actions as an eventing mechanism, dispatching an action either side of the call to `merge` in the save settings request handler:

https://github.com/google/site-kit-wp/blob/db0c22626212552d025da373114844a9e5f6cc3a/includes/Core/Modules/Modules.php#L1129

```php
do_action( 'googlesitekit_pre_save_settings_' . $slug );

$module->get_settings()->merge( (array) $request['data'] );

do_action( 'googlesitekit_save_settings_' . $slug );
```

Then in the `Thank_With_Google` module, we could add hooks for the actions for the module. Making use of an instance variable to hold the pre-save connected state and setting the transient if it changes post-save:

```php
add_action(
	'googlesitekit_pre_save_settings_' . MODULE_SLUG,
	function() {
		$this->pre_update_is_connected = $this->is_connected();
	}
);

add_action(
	'googlesitekit_save_settings_' . MODULE_SLUG,
	function() {
		if ( ! $this->pre_update_is_connected && $this->is_connected() ) {
			// This could be wrapped up in an `on_connected` method, or just inlined like so:
			$this->transients->set( self::TRANSIENT_SETUP_TIMER, time(), WEEK_IN_SECONDS );
		}
	}
);
```

We could conceivably extract and generify this further into a reusable Trait, although I am slightly hesitant on that front because `is_connected` is so free-form there's no guarantee that it will only relate to the settings for a given module. For example the Tag Manager `is_connected` also relies on the current AMP mode.

Anyhow, we could certainly start off with the above. Or, the `SetupCustomize` could still be worth considering, I am leaning toward not doing that though because it would be nice not to have to rely on another HTTP request to set the connected state (say the request failed due to the network, browser window being closed etc., it's adding an extra potential point of failure).

Please see what you think, interested to hear your thoughts.

**Comment by hussain-t** [comment]
Thanks for your suggestion, @techanvil. SGTM!

**Comment by mohitwp** [comment]
### QA Update ❌ 


I found the below issue when testing on site without SSL for the #5806 and also posted my observation in [comment](https://github.com/google/site-kit-wp/issues/5806#issuecomment-1263187236)

As discussed @hussain-t will work on below listed issue and on other observations mentioned in the [comment](https://github.com/google/site-kit-wp/issues/5538#issuecomment-1262342041)
**Issue -** 

If site don't have SSL and if 'TwG' feature flag is active then after completing set up of site when user navigates to dashboard it shows 'map select' error for 'getSupporterWallPrompt'). I verified this with both Twenty twenty and twenty twenty two theme. I'm getting the same error and admin user is not able to access the site kit dashboard due to this error.
Is this error coming due to "Supporter wall widget banner" ?

```
``An error occurred while running 'mapSelect': Cannot read properties of null (reading 'getSupporterWallPrompt')
The error may be correlated with this previous error:
TypeError: Cannot read properties of null (reading 'getSupporterWallPrompt')
    at Object.current (http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-dashboard-678a55075ebc683c20ea.js:74:8776)
    at e (http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-2b9feebf5aa70af8ce1d.js:7:94087)
    at http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-2b9feebf5aa70af8ce1d.js:7:94252
    at http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-2b9feebf5aa70af8ce1d.js:7:83865
    at Array.forEach (<anonymous>)
    at o (http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-2b9feebf5aa70af8ce1d.js:7:83837)
    at http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-2b9feebf5aa70af8ce1d.js:7:82254
    at p (http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-2b9feebf5aa70af8ce1d.js:7:74694)
    at http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-2b9feebf5aa70af8ce1d.js:7:79875
    at http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-2b9feebf5aa70af8ce1d.js:7:77290

Original stack trace:

    in ThankWithGoogleSupporterWallNotification
    in BannerNotifications
    in div
    in Header
    in DashboardMainApp
    in DashboardEntryPoint
    in RestoreSnapshots
    in ErrorHandler
    in ThemeProvider
    in StrictMode
    in Root``
```

<details>

Error on Site kit Dashboard -

![image](https://user-images.githubusercontent.com/94359491/193210085-62615d87-5f42-4274-803f-8532d86f2efd.png)

**Comment by hussain-t** [comment]
Thanks, @mohitwp, regarding your observations:

I will create a follow-up PR for this error:
>An error occurred while running 'mapSelect': Cannot read properties of null (reading 'getSupporterWallPrompt')

----------

>Question 2 > If any banner is already present then on navigating to dashboard for a few milliseconds that banner displays and after that Support wall widget banner appears. Can we do some changes so that other banner not appear when user navigates to dashboard .
The question arises here what is the priority of Supporter Wall Widget banner. Is it have higher priority than zero data states banner notifications ?

The delay is presumably caused by the API request triggered by `getSupporterWallPrompt`. We could consider preloading this API call, or we could move the banner below the zero states notification banner. At the same time, we need to answer the question about "whether the GA4 banner should be a higher priority?". IMHO, this could be a separate issue.

As for the first question, I would like to hear from @aaemnnosttv how to deal if the full site editing is enabled.

**Comment by aaemnnosttv** [comment]
⚠️ After backtracking a regression flagged by @wpdarren, there is an issue introduced as part of this issue where the `supporter-wall-prompt` is being requested any time the TwG feature flag is enabled, regardless if the module is active or not, which was a detail not included in the AC here but one that was probably assumed. I'll pick this up in the review of #5961

**Comment by hussain-t** [comment]
Thanks, @aaemnnosttv. I have already fixed this issue in #5961

**Comment by mohitwp** [comment]
@aaemnnosttv 

Me and @hussain-t need your suggestions regarding the observation I posted earlier. 

**Question -**
I've noticed that supported wall widget reminder banner notification showing also when site have full site editing theme. Currently we can't add supporter wall widget on full site editing theme. Are we going to show banner in this case also ? Currently, we don't have any check for full site editing theme. User can connect TwG module when Full site editing theme enabled and have option to add supporter wall under TwG settings also.

cc @wpdarren

**Comment by aaemnnosttv** [comment]
@mohitwp back to you for another pass 👍

**Comment by mohitwp** [comment]
### QA Update ❌ 

@hussain-t 

- Verified on main and develop branch.
- Error which I reported [here](https://github.com/google/site-kit-wp/issues/5538#issuecomment-1263322636) is still showing on site without SSL. Only difference is that previously I was unable to access Site Kit Dashboard previously now I can access after refreshing the page. Error coming only when TwG module is active and user connects analytics module.

```
``An error occurred while running 'mapSelect': Cannot read properties of null (reading 'getPublicationID')
The error may be correlated with this previous error:
TypeError: Cannot read properties of null (reading 'getPublicationID')
    at Object.current (http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-dashboard-b54aae02bf412137dae6.js:68:14179)
    at e (http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-26b0b943bef4e536710f.js:7:94087)
    at http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-26b0b943bef4e536710f.js:7:94252
    at http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-26b0b943bef4e536710f.js:7:83865
    at Array.forEach (<anonymous>)
    at o (http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-26b0b943bef4e536710f.js:7:83837)
    at http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-26b0b943bef4e536710f.js:7:82254
    at Object.p [as dispatch] (http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-26b0b943bef4e536710f.js:7:74694)
    at e (<anonymous>:2494:20)
    at http://hivevine.com/wp-content/plugins/google-site-kit/dist/assets/js/googlesitekit-vendor-26b0b943bef4e536710f.js:7:79875

Original stack trace:

    in SetupSuccessBannerNotification
    in BannerNotifications
    in div
    in Header
    in DashboardMainApp
    in DashboardEntryPoint
    in RestoreSnapshots
    in ErrorHandler
    in ThemeProvider
    in StrictMode
    in Root``
```



<details>

https://user-images.githubusercontent.com/94359491/194333281-d0406fb1-44bd-4e4c-8a8b-bf4d64b6c486.mp4

</details>

**Apart from that I have 2 questions -**

**1)** I've found that in figma design "Add Supporter Wall Widget" button and may be letter button font is "Google Sans Display" and implemented font is "Google Sans Text". It's match with other button on site. But want to confirm.

<details>

![image](https://user-images.githubusercontent.com/94359491/194334576-f21184ec-c12b-4b54-92d8-991193dbc709.png)

</details>

**2)** Supporter wall widget reminder banner notification currently appears only on main dashboard. Are we not going to show this on entity dashboard ?


**Apart from that I've one suggestion** - When we click on 'Add Supporter Wall Widget" button banner gets disappear . Instead of this can we show loader with button just like we implemented for GA4 banner. 

<details>


https://user-images.githubusercontent.com/94359491/194336351-46777244-a646-4dbc-9a4a-ab59897bf8f9.mp4

</details>

**Comment by hussain-t** [comment]
@mohitwp, the error is not part of the implementation of this ticket, as we confirmed that it happens with the latest release as well. You can create a separate ticket for that.

>1) I've found that in figma design "Add Supporter Wall Widget" button and may be letter button font is "Google Sans Display" and implemented font is "Google Sans Text". It's match with other button on site. But want to confirm.

The implementation is correct. It's the same as the other notifications. All the notification use the same font.

>2) Supporter wall widget reminder banner notification currently appears only on main dashboard. Are we not going to show this on entity dashboard ?

The main dashboard is the most visible place to put it, so I don't think anyone will miss it just because it isn't also on the entity dashboard - as @aaemnnosttv mentioned in this [comment](https://github.com/google/site-kit-wp/issues/5873#issuecomment-1270079090).

>Apart from that I've one suggestion - When we click on 'Add Supporter Wall Widget" button banner gets disappear . Instead of this can we show loader with button just like we implemented for GA4 banner.

Similarly, please refer to this[comment](https://github.com/google/site-kit-wp/issues/5873#issuecomment-1270079090).

However, I want to hear your thoughts on the above, @aaemnnosttv

**Comment by aaemnnosttv** [comment]
> > Apart from that I've one suggestion - When we click on 'Add Supporter Wall Widget" button banner gets disappear . Instead of this can we show loader with button just like we implemented for GA4 banner.
> 
> Similarly, please refer to this[comment](https://github.com/google/site-kit-wp/issues/5873#issuecomment-1270079090).

We already have functionality in place to generally support a spinner for the button when it triggers a background process, however in this case it's just a link to the widgets management screen, so it's a bit different because there's a navigation involved. We can explore this in a new issue as mentioned on the other, but it need not block things here 👍

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Verified on main branch.
- Verified the this [story](https://google.github.io/site-kit-wp/storybook/develop/?path=/story/components-thankwithgooglesupporterwallnotification--ready).
- Verified the new banner notification it is appearing as per the AC.
- Issue mentioned [here](https://github.com/google/site-kit-wp/issues/5538#issuecomment-1263322636) is now resolved. Still errors are showing but those are not due to 'Supported wall widget Reminder' and also exist on latest so I will create separate ticket for this.
- As discussed I will create separate ticket for my observations [here](https://github.com/google/site-kit-wp/issues/5538#issuecomment-1270204900)

<details>

![image](https://user-images.githubusercontent.com/94359491/194500633-0253a152-b0a8-4606-bbfe-c6ddbd4fdd56.png)

</details>
